tests/src/Interop/COM/NativeServer/StringTesting.h

```diff
@@ -0,0 +1,327 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#pragma once
+
+#include "Servers.h"
+
+class DECLSPEC_UUID("C73C83E8-51A2-47F8-9B5C-4284458E47A6") StringTesting : public UnknownImpl, public IStringTesting
+{
+private:
+    template <typename STRING>
+    bool EqualValue(STRING l, STRING r)
+    {
+        STRING tmp = l;
+        int len = 1; // Include 1 for null
+        while (*tmp++)
+            ++len;
+
+        return (0 == ::memcmp(l, r, len * sizeof(l[0])));
+    }
+
+    template <typename STRING>
+    STRING ReverseInplace(size_t len, STRING s)
+    {
+        std::reverse(s, s + len);
+        return s;
+    }
+
+    template<typename STRING>
+    HRESULT Reverse(STRING str, STRING *res)
+    {
+        STRING tmp = str;
+        size_t len = 0;
+        while (*tmp++)
+            ++len;
+
+        size_t strDataLen = (len + 1) * sizeof(str[0]);
+        auto resLocal = (STRING)::CoTaskMemAlloc(strDataLen);
+        if (resLocal == nullptr)
+            return E_OUTOFMEMORY;
+
+        ::memcpy_s(resLocal, strDataLen, str, strDataLen);
+        *res = ReverseInplace(len, resLocal);
+
+        return S_OK;
+    }
+
+    HRESULT ReverseBstr(BSTR str, BSTR *res)
+    {
+        UINT len = ::SysStringLen(str);
+        UINT strDataLen = len * sizeof(str[0]);
+        BSTR resLocal = ::SysAllocStringByteLen(reinterpret_cast<LPCSTR>(str), strDataLen);
+        if (resLocal == nullptr)
+            return E_OUTOFMEMORY;
+
+        *res = ReverseInplace(len, resLocal);
+
+        return S_OK;
+    }
+
+public: // IStringTesting
+    DEF_RAWFUNC(Add_LPStr)(
+        /*[in]*/ LPSTR a,
+        /*[in]*/ LPSTR b,
+        /*[out,retval]*/ LPSTR * pRetVal)
+    {
+        if (a == nullptr || b == nullptr)
+            return E_POINTER;
+
+        size_t aLen = ::strlen(a);
+        size_t bLen = ::strlen(b);
+        auto buf = (LPSTR)::CoTaskMemAlloc((aLen + bLen + 1) * sizeof(*b));
+
+        ::strcpy_s(buf, aLen + 1, a);
+        ::strcpy_s(buf + aLen, bLen + 1, b);
+
+        *pRetVal = buf;
```

Comment:
free a , b.. also other related [in] only scenario

Reply:
I have to free the `LPSTR` `in` parameters?

Reply:
See above